# A complete profile for sentiment analysis model

## Models Ideas:
1. Use of custom models for different categories (tech, food, books,...) to be automatically (using context classfication) or manually selected(by the client). *(different datasets applied)*
2. Run multiple models per dataset and derive weighted average results.
3. Developing a layered classification **use *fast/slow* classification** (divide the dataset using confidence index to strong and weak groups; the weak group will be analysed further using Roberta model).
4. Aspect based analysis **(attach sentiment to specific aspects rather than sentence/opinion)** and word cloud **(for word frequencies)** to show insights of the reviews. (Amazon comprehend model)
5. Use of lemmatization, opinion unit extractor, subjectivity index and multiclass classification(love, sad, angry,...) for better accuracy and data enrichment.
6. Test of a sent-ngrams lexion sentiment analysis **(SO-CAL)**.
7. Use of client dataset to fine-tune the model. (Ideation phase)

## Datasets used:
1. Twitter airline 
2. IMDB 
3. Yelp (preprocessing phase)
4. Amazon 
5. 140sentiment twitter

## Implementation:

### Imports:

In [3]:
# Install the transformers library
!pip install transformers
!pip install vaderSentiment
!pip install flair

In [4]:
from textblob import TextBlob as tb
from textblob.sentiments import NaiveBayesAnalyzer
from textblob import Blobber
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from flair.data import Sentence
from flair.models import TextClassifier
from sklearn import metrics
import pandas as pd
from transformers import pipeline
from sklearn.utils import shuffle
import time
# from transformers import AutoTokenizer, AutoModelForSequenceClassification, Trainer
# only for colab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### libraries implementation:

In [5]:
def metric(true,predict):
    analytics=[]
    #metrics.classification_report(true,predict)
    analytics.append(round(metrics.accuracy_score(true,predict),2))
    analytics.append(round(metrics.precision_score(true,predict,average='weighted'),2))
    analytics.append(round(metrics.recall_score(true,predict,average='weighted'),2))
    analytics.append(round(metrics.f1_score(true,predict,average='weighted'),2))
    return analytics

In [6]:
# class modelDataset:
#     def __init__(self, tokenized_texts):
#         self.tokenized_texts = tokenized_texts

#     def __len__(self):
#         return len(self.tokenized_texts["input_ids"])

#     def __getitem__(self, idx):
#         return {k: v[idx] for k, v in self.tokenized_texts.items()}

In [7]:
def textblobPattern(text):
    sentiment = []
    start=time.time()
    for sentence in text:
        sent = tb(sentence).polarity
        if sent > 0:
            sentiment.append(1)
        elif sent < 0:
            sentiment.append(-1)
        else:
            sentiment.append(0)
    end=time.time()
    tm=(end-start)/len(text)
    return tm,sentiment


In [8]:
def textblobNB(text):
    sentimentV, sentimentT = [], []
    tbnb = Blobber(analyzer=NaiveBayesAnalyzer())
    for sentence in text:
        ts = tbnb(sentence).sentiment
    return sentimentV, sentimentT


In [9]:
def vader(text):
    sentiment = []
    analyzer = SentimentIntensityAnalyzer()
    start=time.time()
    for sentence in text:
        vs = analyzer.polarity_scores(sentence)['compound']
        if (vs > 0.5):
            sentiment.append(1)
        elif (vs < -0.5):
            sentiment.append(-1)
        else:
            sentiment.append(0)
    end = time.time()
    tm = (end - start) / len(text)
    return tm, sentiment


In [10]:
def flair(text):
    classifier = TextClassifier.load('sentiment-fast')
    # sentences = [Sentence(t) for t in text]
    sentiment = []
    start=time.time()
    for phrase in text:
        sentence = Sentence(phrase)
        classifier.predict(sentence, mini_batch_size=32)
        sentiment.append(1 if sentence.labels[0].value == 'POSITIVE' else -1)
    end = time.time()
    tm = (end - start) / len(text)
    return tm, sentiment


In [11]:
def RoBerta_large(text):
    # tokenizer = AutoTokenizer.from_pretrained("siebert/sentiment-roberta-large-english")
    # model = AutoModelForSequenceClassification.from_pretrained("siebert/sentiment-roberta-large-english")
    # trainer = Trainer(model=model)
    # tokenized_texts = tokenizer(text,truncation=True,padding=True)
    # pred_dataset = modelDataset(tokenized_texts)
    # predictions = trainer.predict(pred_dataset)
    # preds = predictions.predictions.argmax(-1)
    # return [-1 if x==0 else 1 for x in preds]
    sentiment = []
    sentiment_analysis = pipeline("sentiment-analysis", model="siebert/sentiment-roberta-large-english")
    start=time.time()
    for x in text:
        sentiment.append(1 if sentiment_analysis(x, truncation=True, padding=True)[0]['label'] == 'POSITIVE' else -1)
    end = time.time()
    tm = (end - start) / len(text)
    return tm, sentiment


In [12]:

def bert_base(text):
    sentiment = []
    sentiment_analysis = pipeline("sentiment-analysis", model="nlptown/bert-base-multilingual-uncased-sentiment")
    start=time.time()
    for x in text:
        sent=  (sentiment_analysis(x, truncation=True, padding=True)[0]['label'])
        if sent== "1 star":
                sentiment.append(-1)
        elif sent== "2 stars":
                sentiment.append(-1)
        elif sent == "3 stars":
                sentiment.append(0)
        elif sent =="4 stars":
                sentiment.append(1)
        elif sent =="5 stars":
                sentiment.append(1)
    end = time.time()
    tm = (end - start) / len(text)
    return tm, sentiment


### Model testing: (for each dataset)

In [13]:
def UsAirTw():
    file_name = "Datasets/usAir_tweets.csv"
    text_column = "text"
    df = pd.read_csv(file_name)
    df=df[:10000]
    true = df["sentiment"]
    pred_texts = df[text_column].dropna().astype('str').tolist()
    print("Metrics for usairline_tweets Dataset:")
    time, pred = textblobPattern(pred_texts)
    textblobM = metric(true, pred)
    print(f"The textblob metrics:\n accuracy={textblobM[0]},precision={textblobM[1]},recall={textblobM[2]},f1 score ={textblobM[3]},time per sentiment= {time}")
    time, pred = vader(pred_texts)
    vaderM = metric(true, pred)
    print(f"The Vader metrics:\n accuracy={vaderM[0]},precision={vaderM[1]},recall={vaderM[2]},f1 score ={vaderM[3]},time per sentiment= {time}")
    time, pred = flair(pred_texts)
    flairM = metric(true, pred)
    print(f"The flair metrics:\n accuracy={flairM[0]},precision={flairM[1]},recall={flairM[2]},f1 score ={flairM[3]},time per sentiment= {time}")
    time, pred = RoBerta_large(pred_texts)
    robertaM = metric(true, pred)
    print(f"The Roberta large metrics:\n accuracy={robertaM[0]},precision={robertaM[1]},recall={robertaM[2]},f1 score ={robertaM[3]},time per sentiment= {time}")
    time, pred = bert_base(pred_texts)
    Bert = metric(true, pred)
    print(f"The bert multilang metrics:\n accuracy={Bert[0]},precision={Bert[1]},recall={Bert[2]},f1 score ={Bert[3]},time per sentiment= {time}")

In [14]:

def yelp():
    file_name = "Datasets/yelp.csv"
    text_column = "text"
    df = pd.read_csv(file_name)
    true = df["sent"]
    pred_texts = df[text_column].dropna().astype('str').tolist()
    print("Metrics for yelp sentiment:")
    time, pred = textblobPattern(pred_texts)
    textblobM = metric(true, pred)
    print(f"The textblob metrics:\n accuracy={textblobM[0]},precision={textblobM[1]},recall={textblobM[2]},f1 score ={textblobM[3]},time per sentiment= {time}")
    time, pred = vader(pred_texts)
    vaderM = metric(true, pred)
    print(f"The Vader metrics:\n accuracy={vaderM[0]},precision={vaderM[1]},recall={vaderM[2]},f1 score ={vaderM[3]},time per sentiment= {time}")
    time, pred = flair(pred_texts)
    flairM = metric(true, pred)
    print(f"The flair metrics:\n accuracy={flairM[0]},precision={flairM[1]},recall={flairM[2]},f1 score ={flairM[3]},time per sentiment= {time}")
    time, pred = RoBerta_large(pred_texts)
    robertaM = metric(true, pred)
    print(f"The Roberta large metrics:\n accuracy={robertaM[0]},precision={robertaM[1]},recall={robertaM[2]},f1 score ={robertaM[3]},time per sentiment= {time}")
    time, pred = bert_base(pred_texts)
    Bert = metric(true, pred)
    print(f"The bert multilang metrics:\n accuracy={Bert[0]},precision={Bert[1]},recall={Bert[2]},f1 score ={Bert[3]},time per sentiment= {time}")

In [15]:
def Imdb():
    file_name = "Datasets/Imdb.csv"
    text_column = "review"
    df = pd.read_csv(file_name)
    df=df[:10000]
    true = df["sentiment"]
    pred_texts = df[text_column].dropna().astype('str').tolist()
    print("Metrics for IMDB dataset:")
    time, pred = textblobPattern(pred_texts)
    textblobM = metric(true, pred)
    print(f"The textblob metrics:\n accuracy={textblobM[0]},precision={textblobM[1]},recall={textblobM[2]},f1 score ={textblobM[3]},time per sentimen= t{time}")
    time, pred = vader(pred_texts)
    vaderM = metric(true, pred)
    print(f"The Vader metrics:\n accuracy={vaderM[0]},precision={vaderM[1]},recall={vaderM[2]},f1 score ={vaderM[3]},time per sentiment= {time}")
    time, pred = flair(pred_texts)
    flairM = metric(true, pred)
    print(f"The flair metrics:\n accuracy={flairM[0]},precision={flairM[1]},recall={flairM[2]},f1 score ={flairM[3]},time per sentiment= {time}")
    time, pred = RoBerta_large(pred_texts)
    robertaM = metric(true, pred)
    print(f"The Roberta large metrics:\n accuracy={robertaM[0]},precision={robertaM[1]},recall={robertaM[2]},f1 score ={robertaM[3]},time per sentiment= {time}")
    time, pred = bert_base(pred_texts)
    Bert = metric(true, pred)
    print(f"The bert multilang metrics:\n accuracy={Bert[0]},precision={Bert[1]},recall={Bert[2]},f1 score ={Bert[3]},time per sentiment= {time}")

In [16]:
def t140sentiment():
    file_name = "/content/drive/MyDrive/140sentiment.csv"
    text_column = "text"
    df = pd.read_csv(file_name)
    df=shuffle (df)
    df=df[:10000]
    true = df["sent"]
    pred_texts = df[text_column].dropna().astype('str').tolist()
    print("Metrics for sentiment140 dataset:")
    # time, pred = textblobPattern(pred_texts)
    # textblobM = metric(true, pred)
    # print(f"The textblob metrics:\n accuracy={textblobM[0]},precision={textblobM[1]},recall={textblobM[2]},f1 score ={textblobM[3]},time per sentiment= {time}")
    # time, pred = vader(pred_texts)
    # vaderM = metric(true, pred)
    # print(f"The Vader metrics:\n accuracy={vaderM[0]},precision={vaderM[1]},recall={vaderM[2]},f1 score ={vaderM[3]},time per sentiment= {time}")
    # time, pred = flair(pred_texts)
    # flairM = metric(true, pred)
    # print(f"The flair metrics:\n accuracy={flairM[0]},precision={flairM[1]},recall={flairM[2]},f1 score ={flairM[3]},time per sentiment= {time}")
    time, pred = RoBerta_large(pred_texts)
    robertaM = metric(true, pred)
    print(f"The Roberta large metrics:\n accuracy={robertaM[0]},precision={robertaM[1]},recall={robertaM[2]},f1 score ={robertaM[3]},time per sentiment= {time}")
    # time, pred = bert_base(pred_texts)
    # Bert = metric(true, pred)
    # print( f"The bert multilang metrics:\n accuracy={Bert[0]},precision={Bert[1]},recall={Bert[2]},f1 score ={Bert[3]},time per sentiment= {time}")

In [17]:
def amazon():
    file_name = "Datasets/amazonFoods.csv"
    text_column = "Text"
    df = pd.read_csv(file_name)
    df=df[:10000]
    true = df["Score"]
    pred_texts = df[text_column].dropna().astype('str').tolist()
    print("Metrics for Amazon food dataset:")
    time,pred=textblobPattern(pred_texts)
    textblobM = metric(true, pred)
    print(f"The textblob metrics:\n accuracy={textblobM[0]},precision={textblobM[1]},recall={textblobM[2]},f1 score ={textblobM[3]},time per sentiment= {time}")
    time,pred=vader(pred_texts)
    vaderM = metric(true, pred)
    print(f"The Vader metrics:\n accuracy={vaderM[0]},precision={vaderM[1]},recall={vaderM[2]},f1 score ={vaderM[3]},time per sentiment= {time}")
    time,pred=flair(pred_texts)
    flairM = metric(true, pred)
    print(f"The flair metrics:\n accuracy={flairM[0]},precision={flairM[1]},recall={flairM[2]},f1 score ={flairM[3]},time per sentiment= {time}")
    time,pred=RoBerta_large(pred_texts)
    robertaM = metric(true, pred)
    print(f"The Roberta large metrics:\n accuracy={robertaM[0]},precision={robertaM[1]},recall={robertaM[2]},f1 score ={robertaM[3]},time per sentiment= {time}")
    time,pred=bert_base(pred_texts)
    Bert = metric(true, pred)
    print(f"The bert multilang metrics:\n accuracy={Bert[0]},precision={Bert[1]},recall={Bert[2]},f1 score ={Bert[3]},time per sentiment= {time}")




### Main:

In [18]:

t140sentiment()


Metrics for sentiment140 dataset:


Downloading:   0%|          | 0.00/687 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/256 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/798k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/456k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/150 [00:00<?, ?B/s]

The Roberta large metrics:
 accuracy=0.75,precision=0.76,recall=0.75,f1 score =0.75,time per sentiment= 0.4011826564311981


## Results:
### for UsAirlineTweets:
- The textblob metrics:
 - accuracy=0.47, precision=0.68, recall=0.47, f1 score =0.48,time per sentiment = 0.0001
- The Vader metrics:
 - accuracy=0.41, precision=0.73, recall=0.41, f1 score =0.4,time per sentiment = 0.0001
- The flair metrics:
 - accuracy=0.67, precision=0.58, recall=0.67, f1 score =0.61 ,time per sentiment = 0.001
- The Roberta large metrics:
 - accuracy=0.75, precision=0.62, recall=0.75, f1 score =0.67,time per sentiment = 0.0173
- The bert multilang metrics:
 - accuracy=0.66, precision=0.64, recall=0.66, f1 score =0.64,time per sentiment = 0.071
 
### for imdb:
- The textblob metrics:
 - accuracy=0.69,precision=0.76,recall=0.69,f1 score =0.67,time per sentimenT= 0.0015
- The Vader metrics:
 - accuracy=0.64,precision=0.75,recall=0.64,f1 score =0.67,time per sentiment= 0.0064
- The flair metrics:
 - accuracy=0.9,precision=0.9,recall=0.9,f1 score =0.9,time per sentiment= 0.039
- The Roberta large metrics:
 - accuracy=0.96,precision=0.96,recall=0.96,f1 score =0.96,time per sentiment= 1.29
- The bert multilang metrics:
 - accuracy=0.78,precision=0.92,recall=0.78,f1 score =0.85,time per sentiment= 0.59

### for Amazon:
- The textblob metrics:
 - accuracy=0.79,precision=0.73,recall=0.79,f1 score =0.75,time per sentiment= 0.001
- The Vader metrics:
 - accuracy=0.75,precision=0.79,recall=0.75,f1 score =0.75,time per sentiment= 0.001
- The flair metrics:
 - accuracy=0.84,precision=0.82,recall=0.84,f1 score =0.82,time per sentiment= 0.013
- The Roberta large metrics:
 - accuracy=0.89,precision=0.84,recall=0.89,f1 score =0.86,time per sentiment= 0.97
- The bert multilang metrics:
 - accuracy=0.85,precision=0.88,recall=0.85,f1 score =0.86,time per sentiment= 0.29

### for 140sentiment:
- The textblob metrics:
 - accuracy=0.44,precision=0.7,recall=0.44,f1 score =0.52,time per sentiment= 0.00026
- The Vader metrics:
 - accuracy=0.3,precision=0.79,recall=0.3,f1 score =0.41,time per sentiment= 0.0001
- The flair metrics:
 - accuracy=0.7,precision=0.7,recall=0.7,f1 score =0.7,time per sentiment= 0.0031
-The Roberta large metrics:
 - accuracy=0.75,precision=0.76,recall=0.75,f1 score =0.75,time per sentiment= 0.401
- The bert multilang metrics:
 - accuracy=0.63,precision=0.72,recall=0.63,f1 score =0.67,time per sentiment= 0.112
 
### for yelp:
- The textblob metrics:
 - accuracy=0.68,precision=0.78,recall=0.68,f1 score =0.65,time per sentiment= 0.001
- The Vader metrics:
 - accuracy=0.62,precision=0.83,recall=0.62,f1 score =0.63,time per sentiment= 0.002
- The flair metrics:
 - accuracy=0.93,precision=0.93,recall=0.93,f1 score =0.93,time per sentiment= 0.022

## Topics covered:
- Textblob - vader - flair libraries
- Text operations: lemmatization - tokenization - vectorization - wordnet - tagging - n-gram 
- Machine learning concepts: vector space model, k-means clustering,[ Naive Bayes, k-NN, SVM] classifiers, decision tree - random forest - transformers (word2vec and wordtree of stanford).
- Technologies: Jupyter notebook - Google colab
- Dataset handeling: dataset preprocessing
- Sentiment analysis approaches
- Handeling multiple  Deeplearning models: roBERTa - BERT - [GloVe - Fasttext - torchtext]

## References:
- https://neptune.ai/blog/sentiment-analysis-python-textblob-vs-vader-vs-flair
- https://towardsdatascience.com/customer-churn-accuracy-a-4-6-increase-with-feature-engineering-29bcb1b1ee8f (REVIEW)
- https://www.analyticsvidhya.com/blog/2021/01/sentiment-analysis-vader-or-textblob/
- https://pythonprogramming.net/sentiment-analysis-python-textblob-vader/
- https://towardsdatascience.com/sentimental-analysis-using-vader-a3415fef7664
- https://medium.com/geekculture/what-nlp-library-you-should-use-for-your-sentimental-analysis-project-bef6b357a6db
- https://towardsdatascience.com/sentiment-analysis-comparing-3-common-approaches-naive-bayes-lstm-and-vader-ab561f834f89
****
* N-grams rule based model
- https://www.sciencedirect.com/science/article/pii/S095741741830143X
- https://github.com/sfu-discourse-lab/SO-CAL(to be reviewed)
- https://towardsdatascience.com/text-analysis-basics-in-python-443282942ec5
****
- https://towardsdatascience.com/text-classification-with-state-of-the-art-nlp-library-flair-b541d7add21f